source/sets_functions_and_relations.rst

```diff
@@ -1293,4 +550,457 @@ at an appropriate point in the proof.
+    variable  f : α → β
+    variables s t : set α
+
+    -- BEGIN
+    example : s ⊆ f ⁻¹' (f '' s) :=
+    begin
+      intros x xs,
+      show f x ∈ f '' s,
+      use [x, xs]
+    end
+    -- END
+
+We can replace the line ``use [x, xs]`` by
+``apply mem_image_of_mem f xs`` if we want to
+use a theorem specifically designed for that purpose.
+But knowing that the image is defined in terms
+of an existential quantifier is often convenient.
+
+Here is a long list of set-theoretic identities for
+you to enjoy.
+You don't have to do all of them at once;
+do a few of them,
+and set the rest aside for a rainy day.
+
+.. code-block:: lean
+
+    import data.set.function
+
+    open set function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+    variables u v : set β
+
+    -- BEGIN
+    example (h : injective f) : f ⁻¹' (f '' s) ⊆ s :=
+    sorry
+
+    example : f '' (f⁻¹' u) ⊆ u :=
+    sorry
+
+    example (h : surjective f) : u ⊆ f '' (f⁻¹' u) :=
+    sorry
+
+    example (h : s ⊆ t) : f '' s ⊆ f '' t :=
+    sorry
+
+    example (h : u ⊆ v) : f ⁻¹' u ⊆ f ⁻¹' v :=
+    sorry
+
+    example : f ⁻¹' (u ∪ v) = f ⁻¹' u ∪ f ⁻¹' v :=
+    sorry
+
+    example : f '' (s ∩ t) ⊆ f '' s ∩ f '' t :=
+    sorry
+
+    example (h : injective f) : f '' s ∩ f '' t ⊆ f '' (s ∩ t) :=
+    sorry
+
+    example : f '' s \ f '' t ⊆ f '' (s \ t) :=
+    sorry
+
+    example : f ⁻¹' u \ f ⁻¹' v ⊆ f ⁻¹' (u \ v) :=
+    sorry
+
+    example : f '' s ∩ v = f '' (s ∩ f ⁻¹' v) :=
+    sorry
+
+    example : f '' (s ∩ f ⁻¹' u) ⊆ f '' s ∪ u :=
+    sorry
+
+    example : s ∩ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∩ u) :=
+    sorry
+
+    example : s ∪ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∪ u) :=
+    sorry
+    -- END
+
+You might also enjoy looking up for making up some
+identities involving indexed unions and proving them.
+
+The fact that in type theory a function is always totally
+defined on its domain type
+sometimes forces some difficult choices.
+For example, if we want to define ``x / y`` and ``log x``
+as functions on the reals,
+we have to assign a value to the first when ``y`` is ``0``,
+and a value to the second for ``x ≤ 0``.
+The strategy generally followed by the Lean library
+in these situations is to assign such functions somewhat arbitrary
+but convenient values outside their natural domain.
+For example, defining ``x / 0`` to be ``0`` means that the
+identity ``(x + y) / z = x / 2 + y / z`` holds
+for every ``x``, ``y``, and ``z``.
+When you see a theorem in the library that uses the
+division symbol,
+you should be mindful that theorem depends on this
+nonstandard definition,
+but this generally does not cause problems in practice.
+When we need to,
+we can restrict the statement of a theorem so that
+it does not rely on such values.
+For example, if a theorem begins ``∀ x > 0, ...``,
+dividing by ``x`` in the body of the statement is not problematic.
+Limiting the scope of a quantifier in this way is known
+as *relativization*.
```

Comment:
I added the following explanation, together with another `TODO` to come back later and see if we can find a place to discuss them in the logic chapter.

"These are sometimes known as *bounded quantifiers*,
because the construction serves to restrict their significance
to the set ``s``.
As a result, theorems in the library that make use of them
often contain ``ball`` or ``bexists`` in the name."